Web/Edubase.Web.UI/Areas/Groups/Models/Validators/GroupEditorViewModelValidator.cs

```diff
@@ -42,10 +42,32 @@ public GroupEditorViewModelValidator(IGroupReadService groupReadService, IEstabl
                     .WithMessage("This establishment is already in this group. Please enter a different URN")
                     .WithSummaryMessage("This establishment is already in this group. Please enter a different URN")
 
-                    .MustAsync(async (x, ct) =>
+                    .MustAsync(async (urnSearchText, ct) =>
                     {
-                        return (await _establishmentReadService.GetAsync(x.ToInteger().Value, principal).ConfigureAwait(false)).ReturnValue != null;
-                    }).WithMessage("The establishment was not found").WithSummaryMessage("The establishment was not found");
+                        var matchedEstablishment = (await _establishmentReadService.GetAsync(urnSearchText.ToInteger().Value, principal).ConfigureAwait(false)).ReturnValue;
+                        var establishmentExists = matchedEstablishment != null;
+                        return establishmentExists;
+                    })
+                    .WithMessage("The establishment was not found")
+                    .WithSummaryMessage("The establishment was not found")
+
+                    .MustAsync(async (model, urnSearchText, ct) =>
+                    {
+                        var matchedEstablishment = (await _establishmentReadService.GetAsync(urnSearchText.ToInteger().Value, principal).ConfigureAwait(false)).ReturnValue;
```

Comment:
I dislike the duplicated call to the establishment read service, but currently cannot see a way to share state between multiple fluent validation rules, where this value is asynchronously populated.

Ideas/suggestions very much welcomed!

Reply:
Lightbulb moment as soon as submitting this comment(!) 

Marking as resolved.